File: packages/react-native-autoplay/src/templates/NowPlayingTemplate.ts
import { Platform } from 'react-native';
import { NitroModules } from 'react-native-nitro-modules';
import uuid from 'react-native-uuid';
import { HybridAutoPlay } from '..';
import type { NowPlayingTemplate as NitroNowPlayingTemplateSpec } from '../specs/NowPlayingTemplate.nitro';
import type { NitroTemplateConfig, TemplateConfig } from './Template';
import { type NowPlayingButton, NitroNowPlayingButtonUtil } from '../utils/NitroNowPlayingButton';
import type { NitroImage } from '../utils/NitroImage';

const HybridNowPlayingTemplate =
  Platform.OS !== 'web'
    ? NitroModules.createHybridObject<NitroNowPlayingTemplateSpec>('NowPlayingTemplate')
    : null;

export type NitroNowPlayingButtonType =
  | 'shuffle'
  | 'addToLibrary'
  | 'more'
  | 'playbackRate'
  | 'repeat'
  | 'image';

/**
 * Used to convert the very specific typescript typing to an easier to handle type for native code
 */
export type NitroNowPlayingButton = {
  type: NitroNowPlayingButtonType;
  image?: NitroImage;
  onPress: (isSelected: boolean) => void;
  isSelected?: boolean;
};

export interface NitroNowPlayingTemplateConfig extends TemplateConfig {
  buttons?: Array<NitroNowPlayingButton>;
  albumArtistButtonEnabled?: boolean;
  upNextButtonEnabled?: boolean;
  upNextButtonTitle?: string;
  onUpNextButtonPressed?: () => void;
  onAlbumArtistButtonPressed?: () => void;
}

export type NowPlayingTemplateConfig = Omit<NitroNowPlayingTemplateConfig, 'buttons'> & {
  buttons?: Array<NowPlayingButton<NowPlayingTemplate>>;
};

export class NowPlayingTemplate {
  private template = this;
  public id: string;

  constructor(config: NowPlayingTemplateConfig) {
    this.id = uuid.v4();

    if (Platform.OS === 'web') {
      console.warn('NowPlayingTemplate is not supported on web');
      return;
    }

    const { buttons, ...rest } = config;
    const nitroConfig: NitroNowPlayingTemplateConfig & NitroTemplateConfig = {
      ...rest,
      id: this.id,
      buttons: NitroNowPlayingButtonUtil.convert(this.template, buttons),
    };

    HybridNowPlayingTemplate?.configureNowPlayingTemplate(nitroConfig);
  }

  public push(): Promise<void> {
    if (Platform.OS === 'web' || !HybridNowPlayingTemplate) {
      return Promise.reject(new Error('NowPlayingTemplate is not supported on this platform'));
    }
    return HybridAutoPlay.pushTemplate(this.id);
  }

  public updateButtons(buttons: Array<NowPlayingButton<NowPlayingTemplate>>): Promise<void> {
    if (!HybridNowPlayingTemplate) {
      return Promise.reject(new Error('NowPlayingTemplate is not supported on this platform'));
    }
    return HybridNowPlayingTemplate.updateNowPlayingButtons(
      this.id,
      NitroNowPlayingButtonUtil.convert(this.template, buttons) ?? []
    );
  }

  public setAlbumArtistButtonEnabled(enabled: boolean): Promise<void> {
    if (!HybridNowPlayingTemplate) {
      return Promise.reject(new Error('NowPlayingTemplate is not supported on this platform'));
    }
    return HybridNowPlayingTemplate.setAlbumArtistButtonEnabled(this.id, enabled);
  }

  public setUpNextButtonEnabled(enabled: boolean, title?: string): Promise<void> {
    if (!HybridNowPlayingTemplate) {
      return Promise.reject(new Error('NowPlayingTemplate is not supported on this platform'));
    }
    return HybridNowPlayingTemplate.setUpNextButtonEnabled(this.id, enabled, title);
  }
}

File: packages/react-native-autoplay/ios/templates/NowPlayingTemplate.swift
//
//  NowPlayingTemplate.swift
//  react-native-autoplay
//
//  Created by Claude Code
//

import CarPlay

class NowPlayingTemplate: NSObject, AutoPlayTemplate, CPNowPlayingTemplateObserver {
    let template: CPNowPlayingTemplate
    var config: NowPlayingTemplateConfig

    var autoDismissMs: Double? {
        return config.autoDismissMs
    }

    func getTemplate() -> CPTemplate {
        return template
    }

    init(config: NowPlayingTemplateConfig) {
        self.config = config
        self.template = CPNowPlayingTemplate.shared
        super.init()

        // Set up observers for upNext and albumArtist buttons
        template.add(self)

        // Configure initial state
        Task { @MainActor in
            self.invalidate()
        }
    }

    deinit {
        template.remove(self)
    }

    @MainActor
    func invalidate() {
        // Update album artist button
        template.isAlbumArtistButtonEnabled = config.albumArtistButtonEnabled ?? false

        // Update up next button
        template.isUpNextButtonEnabled = config.upNextButtonEnabled ?? false
        if let title = config.upNextButtonTitle {
            template.upNextTitle = title
        }

        // Update now playing buttons
        updateNowPlayingButtons(buttons: config.buttons)
    }

    @MainActor
    private func updateNowPlayingButtons(buttons: [NitroNowPlayingButton]?) {
        guard let buttons = buttons else {
            template.updateNowPlayingButtons([])
            return
        }

        var cpButtons: [CPNowPlayingButton] = []

        for button in buttons {
            let cpButton = createCPButton(from: button)
            if let cpButton = cpButton {
                cpButtons.append(cpButton)
            }
        }

        template.updateNowPlayingButtons(cpButtons)
    }

    @MainActor
    private func createCPButton(from button: NitroNowPlayingButton) -> CPNowPlayingButton? {
        let handler: (CPNowPlayingButton) -> Void = { [weak self] _ in
            button.onPress(button.isSelected ?? false)
        }

        switch button.type {
        case .shuffle:
            let shuffleButton = CPNowPlayingShuffleButton(handler: handler)
            if let isSelected = button.isSelected {
                shuffleButton.isSelected = isSelected
            }
            return shuffleButton

        case .addtolibrary:
            let addButton = CPNowPlayingAddToLibraryButton(handler: handler)
            if let isSelected = button.isSelected {
                addButton.isSelected = isSelected
            }
            return addButton

        case .more:
            return CPNowPlayingMoreButton(handler: handler)

        case .playbackrate:
            return CPNowPlayingPlaybackRateButton(handler: handler)

        case .repeat:
            let repeatButton = CPNowPlayingRepeatButton(handler: handler)
            if let isSelected = button.isSelected {
                repeatButton.isSelected = isSelected
            }
            return repeatButton

        case .image:
            guard let imageData = button.image else {
                return nil
            }
            guard let image = Parser.parseNitroImage(
                image: imageData,
                traitCollection: SceneStore.getRootTraitCollection()
            ) else {
                return nil
            }
            return CPNowPlayingImageButton(image: image, handler: handler)
        }
    }

    func onWillAppear(animated: Bool) {
        config.onWillAppear?(animated)
    }

    func onDidAppear(animated: Bool) {
        config.onDidAppear?(animated)
    }

    func onWillDisappear(animated: Bool) {
        config.onWillDisappear?(animated)
    }

    func onDidDisappear(animated: Bool) {
        config.onDidDisappear?(animated)
    }

    func onPopped() {
        // Clean up observers
        template.remove(self)
        config.onPopped?()
    }

    func traitCollectionDidChange() {
        // NowPlaying template doesn't need trait collection updates
    }

    @MainActor
    func updateButtons(buttons: [NitroNowPlayingButton]) {
        config.buttons = buttons
        updateNowPlayingButtons(buttons: buttons)
    }

    @MainActor
    func setAlbumArtistButtonEnabled(enabled: Bool) {
        config.albumArtistButtonEnabled = enabled
        template.isAlbumArtistButtonEnabled = enabled
    }

    @MainActor
    func setUpNextButtonEnabled(enabled: Bool, title: String?) {
        config.upNextButtonEnabled = enabled
        template.isUpNextButtonEnabled = enabled
        if let title = title {
            config.upNextButtonTitle = title
            template.upNextTitle = title
        }
    }

    // MARK: - CPNowPlayingTemplateObserver

    func nowPlayingTemplateUpNextButtonTapped(_ nowPlayingTemplate: CPNowPlayingTemplate) {
        config.onUpNextButtonPressed?()
    }

    func nowPlayingTemplateAlbumArtistButtonTapped(_ nowPlayingTemplate: CPNowPlayingTemplate) {
        config.onAlbumArtistButtonPressed?()
    }
}
